pkg/rpcmiddleware/rpcretry/client_interceptors.go

```diff
@@ -0,0 +1,183 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcretry
+
+import (
+	"context"
+	"math/rand"
+	"strconv"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/backoffutils"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	headerXRateLimit     = "x-rate-limit-limit"
+	headerXRateAvailable = "x-rate-limit-available"
+	headerXRateReset     = "x-rate-limit-reset"
+	headerXRateRetry     = "x-rate-limit-retry"
+)
+
+type attemptMetadataKeyType struct{}
+
+var attemptMetadataKey attemptMetadataKeyType
+var errInvalidStreamType = errors.DefineUnavailable("unsupported_stream_type_grpcretry", "Client streams are not supported in the gRPC retry interceptor")
+
+// UnaryClientInterceptor returns a new unary client interceptor that retries the execution of external gRPC calls, the
+// retry attempt will only occur if any of the validators define the error as a trigger.
+func UnaryClientInterceptor(opts ...Option) grpc.UnaryClientInterceptor {
+	callOpts := evaluateClientOpt(opts...)
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		var md metadata.MD
+		var err error
+		logger := log.FromContext(ctx)
+
+		for attempt := uint(0); attempt <= callOpts.max; attempt++ {
+			retryTimeout := callOpts.timeout
+			if callOpts.enableMetadata {
+				if headerTimeout := getHeaderLimiterTimeout(ctx, md, callOpts); headerTimeout > 0 {
+					retryTimeout = headerTimeout
+				}
+			}
+
+			err = waitRetryBackoff(ctx, retryTimeout, attempt)
+			if err != nil {
+				logger.WithError(err).Debug("An unexpected error occurred while in the timeout for the next request retry")
+				return err
+			}
+
+			callCtx := context.WithValue(ctx, attemptMetadataKey, attempt)
+			err = invoker(callCtx, method, req, reply, cc, append(opts, grpc.Header(&md))...)
+			if err == nil {
+				return nil
+			}
+
+			if !isRetriable(err, callOpts) {
+				return err
+			}
+		}
+
+		return err
+	}
+}
+
+// StreamClientInterceptor returns a new streaming client interceptor that retries the execution of external gRPC
+// calls, the retry attempt will only occur if any of the validators define the error as a trigger.
+func StreamClientInterceptor(opts ...Option) grpc.StreamClientInterceptor {
+	callOpts := evaluateClientOpt(opts...)
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		if callOpts.max == 0 {
+			return streamer(ctx, desc, cc, method, opts...)
+		}
+		if desc.ClientStreams {
+			return nil, errInvalidStreamType.New()
+		}
```

Comment:
It's way too easy to break a client this way. If the interceptor doesn't support certain stream types, just pass them onto the next interceptor.

```suggestion
		if desc.ClientStreams || callOpts.max == 0 {
			return streamer(ctx, desc, cc, method, opts...)
		}
```